### PR TITLE
[4.0][com_finder] move finder menu items from sidebar to menu

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
Moved menu items for smart search below the "Smart Search" menu item in the admin menu and removed the sidebar items

### Expected result
Screenshot from 2019-11-09 11-45-27

### CODE REVIEW

**Comment by brianteeman** [comment]
Didn't I already do this ages ago

**Comment by alikon** [comment]
yes #26137 or #24020 
but you have used sql script for menu table which is  **_dangerous_** for nested set tables like `#__menu`

**Comment by richard67** [comment]
Drone failure seems not to be related to this PR.

**Comment by richard67** [comment]
@alikon As I can see, @brianteeman has closed his PR #26137, and PR #24020 is by @Hackwar .

**Comment by Quy** [comment]
No submenu shown using your branch and dev branch with Patchtester.

File: administrator/components/com_finder/View/Filters/HtmlView.php [hunk 1]
```diff
@@ -107,8 +98,6 @@ public function display($tpl = null)
 		$this->filterForm    = $this->get('FilterForm');
 		$this->activeFilters = $this->get('ActiveFilters');
 
-		FinderHelper::addSubmenu('filters');
```
Comment (Quy) [inline administrator/components/com_finder/View/Filters/HtmlView.php:109]: Remove `use Joomla\Component\Finder\Administrator\Helper\FinderHelper;`

File: administrator/components/com_finder/View/Maps/HtmlView.php [hunk 2]
```diff
@@ -111,8 +102,6 @@ public function display($tpl = null)
 		$this->filterForm    = $this->get('FilterForm');
 		$this->activeFilters = $this->get('ActiveFilters');
 
-		FinderHelper::addSubmenu('maps');
```
Comment (Quy) [inline administrator/components/com_finder/View/Maps/HtmlView.php:113]: Remove `use Joomla\Component\Finder\Administrator\Helper\FinderHelper;`

File: administrator/components/com_finder/View/Searches/HtmlView.php [hunk 3]
```diff
@@ -102,8 +102,6 @@ public function display($tpl = null)
 			throw new GenericDataException(implode("\n", $errors), 500);
 		}
 
-		FinderHelper::addSubmenu('searches');
```
Comment (Quy) [inline administrator/components/com_finder/View/Searches/HtmlView.php:105]: Remove `use Joomla\Component\Finder\Administrator\Helper\FinderHelper;`

File: administrator/components/com_finder/Helper/FinderHelper.php [hunk 4]
```diff
@@ -29,39 +28,6 @@ class FinderHelper
...
-	 * @param   string  $vName  The name of the active view.
-	 *
-	 * @return  void
-	 *
-	 * @since   2.5
-	 */
-	public static function addSubmenu($vName)
-	{
-		\JHtmlSidebar::addEntry(
-			Text::_('COM_FINDER_SUBMENU_INDEX'),
```
Comment (Quy) [inline administrator/components/com_finder/Helper/FinderHelper.php:44]: Remove language strings.

**Comment by alikon** [comment]
to test on a fresh install you should comment  those lines https://github.com/joomla/joomla-cms/blob/23bf6185a35aa44c0033c7018d45772d4d7ddc5f/administrator/components/com_admin/script.php#L6092-L6100 otherwise you should test the update from 3.10 having an install with this patch

File: administrator/components/com_finder/View/Filters/HtmlView.php (hunk 1 above)
Comment (alikon) [reply administrator/components/com_finder/View/Filters/HtmlView.php:109]: done

File: administrator/components/com_finder/View/Maps/HtmlView.php (hunk 2 above)
Comment (alikon) [reply administrator/components/com_finder/View/Maps/HtmlView.php:113]: done

File: administrator/components/com_admin/script.php [hunk 5]
```diff
@@ -6104,6 +6104,68 @@ public function postflight($action, $installer)
...
+				'parent_id' => $menuItem['parent_id'],
+				'client_id' => $menuItem['client_id'],
+			];
+
+			if ($tableItem->load($keys))
+			{
+				continue;
+			}
+
+			$newTableItem   = new \Joomla\Component\Menus\Administrator\Table\MenuTable($db);
```
Comment (Quy) [inline administrator/components/com_admin/script.php]:
```suggestion
			$newTableItem = new \Joomla\Component\Menus\Administrator\Table\MenuTable($db);
```

File: administrator/components/com_finder/View/Searches/HtmlView.php (hunk 3 above)
Comment (alikon) [reply administrator/components/com_finder/View/Searches/HtmlView.php:105]: done

File: administrator/components/com_admin/script.php [hunk 6]
```diff
@@ -6202,51 +6264,166 @@ public function postflight($action, $installer)
...
+		$parentId    = $tableItem->id;
+		$componentId = ExtensionHelper::getExtensionRecord('com_finder')->extension_id;
 
-			// Rebuild the tree path.
-			if (!$newTableItem->rebuildPath($newTableItem->id))
-			{
-				return false;
-			}
-		}
+		// Add Contact Fields Menu Items.
```
Comment (Quy) [inline administrator/components/com_admin/script.php]:
```suggestion
		// Add Finder Fields Menu Items.
```

File: administrator/components/com_finder/Helper/FinderHelper.php (hunk 4 above)
Comment (alikon) [reply administrator/components/com_finder/Helper/FinderHelper.php:44]: done

**Comment by Quy** [comment]
Commented the lines and still no go. Shouldn't it work with a new install of v4 without having to comment out these lines?

When installing the patch on v3.10:
> The file marked for modification does not exist: administrator/components/com_finder/Helper/FinderHelper.php

**Comment by alikon** [comment]
forgive me i've completely forgot to provide the sql for the install :face_with_head_bandage: 
and told you wrong path..sorry

**Comment by richard67** [comment]
And what about Joomla.sql for PostgreSQL?

**Comment by alikon** [comment]
added

**Comment by Quy** [comment]
Add language strings.

finder

**Comment by alikon** [comment]
just refresh the page string are there

**Comment by Quy** [comment]
Refreshed. Still an issue.

**Comment by Quy** [comment]
See the other PR for the new language strings.

https://github.com/joomla/joomla-cms/pull/26137/files#diff-a37aa6c91545033de1ec16c00cf31733

**Comment by alikon** [comment]
@Quy is not matter of missing string.... visit another menu item and return back to the  smart search menu item.... strings suddenly are translated

**Comment by Quy** [comment]
@alikon Sorry you are right. Thanks.

**Comment by alikon** [comment]
no problem and thanks for your patience

**Comment by Quy** [comment]
Actually, it is still an issue. 
Go to Newsfeeds > Feed. 
Click Smart Search. See strings not translated.

**Comment by alikon** [comment]
that's really weird...  i confirm your findings .... but no clue

**Comment by Quy** [comment]
I don't know either, however, adding the new language strings fixes the issue.

**Comment by alikon** [comment]
you are right....

File: installation/sql/mysql/joomla.sql [hunk 7]
```diff
@@ -1352,24 +1352,29 @@ SELECT 12, 'main', 'com_newsfeeds', 'News Feeds', '', 'News Feeds', 'index.php?o
...
-
-
-
+SELECT 21, 'main', 'com_finder_index', 'Smart-Search-Index', '', 'Smart Search/Index', 'index.php?option=com_finder&view=index', 'component', 1, 15, 2, `extension_id`, 0, NULL, 0, 0, 'class:associations', 0, '', 32, 33, 0, '*', 1, NULL, NULL FROM `#__extensions` WHERE `name` = 'com_finder';
```
Comment (Quy) [inline installation/sql/mysql/joomla.sql]: Change `class:associations`. I assume it should be `class:finder`. For the others maybe `class:finder-maps`,  `class:finder-filters`,  `class:finder-searches`?

**Comment by brianteeman** [comment]
Menu strings have to be in the sys.ini as the other language file is only available to the component

File: installation/sql/mysql/joomla.sql (hunk 7 above)
Comment (alikon) [reply installation/sql/mysql/joomla.sql]: done

**Comment by brianteeman** [comment]
How to get the code in com_admin/script.php to run?

**Comment by alikon** [comment]
i'm afraid you need to create a joomla pkg from this branch, and then update from 3.10

**Comment by richard67** [comment]
You can find an update package based on nightly build of tonight + this PR applied here:
[https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-12-08_pr-27032.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-12-08_pr-27032.zip)

To test the update, download the package and use the "Upload&Update" tab of the Joomla Update Component. You can test updating a 3.10 or a 3.9.13 or staging in this way, whatever you have.

**Comment by richard67** [comment]
@alikon Postgresql system test fails at installation. Not sure if related to this PR. Have to test here but this can take a while.

**Comment by alikon** [comment]
well i've tested/writed this pr on postgresql...... if you can tell me WTF that failure means i would be happy to fix it.....
....more than often there are this kind of failure... and OK i'm a bad coder i know.... but this happens even from not my pr's....so...

**Comment by richard67** [comment]
Maybe not related to your PR but some other recent change in 4.0-dev, don't know yet. Will check.

**Comment by richard67** [comment]
@alikon I've found it. Will provide fix in a PR to your branch in a few minutes, so you can merge with 1 click.

**Comment by richard67** [comment]
@alikon See [https://github.com/alikon/joomla-cms/pull/62](https://github.com/alikon/joomla-cms/pull/62).

**Comment by richard67** [comment]
The update zip package I've linked above is still good because the last commit only changed postgresql/joomla.sql for new install.

**Comment by richard67** [comment]
@alikon See [https://github.com/alikon/joomla-cms/pull/63](https://github.com/alikon/joomla-cms/pull/63).

**Comment by richard67** [comment]
Now there are conflicts in language files.

**Comment by richard67** [comment]
@alikon Now you have to rename the language files after PR #27130 has been merged.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on d7c03e75de187ae66b7b624c76177a02a9498499<br /><br />I've tested as follows:

Test 1: New installation

1. On a clean current 4.0-dev installation, applied the patch of this PR (e.g. with [Patch Tester 3.0.0 Beta 4](https://github.com/joomla-extensions/patchtester/releases/tag/3.0.0-beta4))
2. Removed configuration.php and deleted database tables.
3. Made a new install.
Result: See section "Expected result" in the description of this PR.

Test 2: Update

1. Updated a clean staging using the "Upload&Update" tab of the Joomla Update Component and the patched update package based on last nightly build + this PR downloaded from here: [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-12-08_pr-27032.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-12-08_pr-27032.zip)
Result: See section "Expected result" in the description of this PR.

I've done both tests with MySQL 5.7 and PostgreSQL 10.10, but for the update test this was not really necessary.

Error reporting was set to "Development", no new PHP errors or warning or notices and nothing bad in the database server's log file added by this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27032">issues.joomla.org/tracker/joomla-cms/27032</a>.</sub>

**Comment by Quy** [comment]
Upgraded from staging and 3.10. No menu items below Smart Search in the side menu.

> PHP Warning:  count(): Parameter must be an array or an object that implements Countable in \plugins\installer\override\override.php on line 134

PHP 7.3.9 MySQLi 10.4.6-MariaDB

**Comment by richard67** [comment]
The error about template is normal when updating to 4.

The warning about SQL error seems to be new, I did not have that when testing => maybe the result of a recent merge into 4.0-dev?

**Comment by richard67** [comment]
The warning aboutr override.php I had, too, but with and without the PR, and it did not do any harm.

**Comment by richard67** [comment]
@Quy What happens if you update to a "normal" nightly build, i.e. not my patched one, i.e. without this PR?

**Comment by richard67** [comment]
I've just tested again and can't replicate @Quy 's issue.

**Comment by Quy** [comment]
Same errors without your patched one.

**Comment by richard67** [comment]
Anything in the log file of the db server? On MySQL and Linux it would be "/var/log/mysql/error.log". No idea how it is with MariaDB.

**Comment by Quy** [comment]
It could be that the upgrade is incomplete for the following columns in menu table:

`checked_out_time | datetime |   |   | No | 0000-00-00 00:00:00`
`publish_up` and `publish_down` missing

**Comment by richard67** [comment]
Yes, that could be. Reason is maybe that you have some sql mode set which we don't support? As far as I remember from a recent posting by @HLeithner , it could be STRICT_ALL_TABLES, while we (Joomla) uses STRICT_TRANS_TABLES.

**Comment by richard67** [comment]
It could be solved by combining all alter table statements for all datetime columns into 1 single alter table statement, so when one columns is altered, there is no other column anymore having the invalid default value '0000 ...', which causes the alter table fail in case of strict mode. But the database schema checker/fixer does not support that and so would show errors after a new install where no errors are, and it would fail trying to fix that. To change the db checker/fixer to support that is one of the things I wanna do when I have much time, but it is not trivial. That's why I think I can't do this before my xmas holiday (if ever).

**Comment by richard67** [comment]
The menu table is the first one which will be altered by a 4.0 update sql, and that's why this table fails, because it comes first.

**Comment by richard67** [comment]
How you can test that:

First run following SQL script in PhpMyAdmin, having replaces `#__` by your db prefix:

```
--
-- Switch off STRICT_TABLES by setting some SQL_MODE without STRICT_TABLES,
-- otherwise you get following SQL error:
-- #1067 - Invalid default value for 'test_col_1'
--

SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";

--
-- Set timezone to UTC so CURRENT_TIMESTAMP is stored as UTC
--
SET time_zone = "+00:00";

--
-- Table structure for table `#__test_table_1`
--

CREATE TABLE IF NOT EXISTS `#__test_table_1` (
  `id` int(10) NOT NULL AUTO_INCREMENT,
  `test_col_1` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
  `test_col_2` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
  `test_col_3` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
  `test_col_4` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
  PRIMARY KEY (`id`),
  KEY `idx_test_col_1` (`test_col_1`),
  KEY `idx_test_col_2` (`test_col_2`),
  KEY `idx_test_col_3` (`test_col_3`),
  KEY `idx_test_col_4` (`test_col_4`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 DEFAULT COLLATE=utf8mb4_unicode_ci;

--
-- Dumping data for table `#__test_table_1`
--

INSERT INTO `#__test_table_1` (`id`, `test_col_1`, `test_col_2`, `test_col_3`, `test_col_4`) VALUES
(1, '0000-00-00 00:00:00', '0000-00-00 00:00:00', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP);

INSERT INTO `#__test_table_1` (`id`, `test_col_1`, `test_col_2`, `test_col_3`, `test_col_4`) VALUES
(2, '2019-01-01 00:00:00', '0000-00-00 00:00:00', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP);

INSERT INTO `#__test_table_1` (`id`, `test_col_1`, `test_col_2`, `test_col_3`, `test_col_4`) VALUES
(3, '0000-00-00 00:00:00', '2019-01-01 00:00:00', CURRENT_TIMESTAMP, CURRENT_TIMESTAMP);
```

Then i a new database session, i.e. new instance of PhpMyAdmin, run the following, also with the right db prefix:

```
ALTER TABLE `#__test_table_1` MODIFY `test_col_1` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP;
ALTER TABLE `#__test_table_1` MODIFY `test_col_2` datetime NOT NULL DEFAULT '1000-01-01 00:00:00';
ALTER TABLE `#__test_table_1` MODIFY `test_col_3` datetime NULL DEFAULT NULL;
ALTER TABLE `#__test_table_1` MODIFY `test_col_4` datetime NULL DEFAULT NULL;
```

This will fail with error invalid default value for datetime column so-and-so (I think it will be `test_col_2`).

Then run following, also with correct db prefix:

```
ALTER TABLE `#__test_table_1`
	MODIFY `test_col_1` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
	MODIFY `test_col_2` datetime NOT NULL DEFAULT '1000-01-01 00:00:00',
	MODIFY `test_col_3` datetime NULL DEFAULT NULL,
	MODIFY `test_col_4` datetime NULL DEFAULT NULL;
```

You will see that the latter works because when the table is altered there will not be any column anymore with invalid default value after the statement has run.

**Comment by Quy** [comment]
> JInstaller: :Install: Error SQL Incorrect datetime value: '' for column 'core_checked_out_time' at row 1

`core_checked_out_time | varchar(255) | utf8mb4_unicode_ci |   | No | 0000-00-00 00:00:00`

Upgraded on CloudAccess. Does it matter that the column is varchar and not datetime initially?

In the following sql, the 4 datetime columns prior to `core_checked_out_time` updated but failed at `core_checked_out_time`.

\administrator\components\com_admin\sql\updates\mysql\4.0.0-2018-08-01.sql
```
ALTER TABLE `#__ucm_content` MODIFY `core_created_time` datetime NOT NULL;
ALTER TABLE `#__ucm_content` MODIFY `core_modified_time` datetime NOT NULL;

ALTER TABLE `#__ucm_content` MODIFY `core_publish_up` datetime NULL DEFAULT NULL;
ALTER TABLE `#__ucm_content` MODIFY `core_publish_down` datetime NULL DEFAULT NULL;
ALTER TABLE `#__ucm_content` MODIFY `core_checked_out_time` datetime NULL DEFAULT NULL;

ALTER TABLE `#__ucm_history` MODIFY `save_date` datetime NOT NULL;
```

**Comment by richard67** [comment]
> Does it matter that the column is varchar and not datetime initially?

Seems so, but I don't see yet why, and I can't reproduce the problem here with a local MySQL 5.7 installation on Linux.

**Comment by Quy** [comment]
CloudAccess system information:
Linux lamp129.cloudaccess.net 3.10.0-962.3.2.lve1.5.24.5.el6h.x86_64
5.7.23-cll-lve

**Comment by richard67** [comment]
@Quy Please execute following 2 SQL statements in PhpMyAdmin or whatever tool and post the result of each of the statements here in some text quote block, if not too big, or send it to me on glip:

```
SELECT @@GLOBAL.sql_mode;
SELECT @@SESSION.sql_mode;
```

I'll check on weekend and if I find some difference to my settings, and if so, I'll apply them on my DB and see if I can replicate the issue.

**Comment by richard67** [comment]
@Quy P.S.: Maybe we should continue this discussion somewhere else because it is not really related to this PR here but to the database update in general?

**Comment by Quy** [comment]
OK. Lets do in Glip.

**Comment by richard67** [comment]
Updated the update package: [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-12-08_pr-27032.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-12-08_pr-27032.zip).

**Comment by richard67** [comment]
Issues with that update package linked above and e.g. untranslated texts in backend are not related to this PR, they happen also with the unmodified update package from tonight's nightly build.

**Comment by Quy** [comment]
Upgraded successfully with no sample data, but failed with sample data.

**Comment by richard67** [comment]
@Quy Which kind of sample data? Blog? Testing? Multilingual? Btw. after update there might be problems with language files not being deleted because deleted files list in script.php is not up to date, but that's another thing not related to this PR, just wanted to mention.

**Comment by Quy** [comment]
Installed `Test English`.

**Comment by richard67** [comment]
@Quy I see. Will test later, too.

**Comment by richard67** [comment]
@Quy Confirmed. But not related to this PR. Will you make a new issue? Unfortunately I see nothing in mysql or php log about the reason, but I can investigate and try to find a fix now as I can reprocude it.

**Comment by richard67** [comment]
@Quy I think I have a solution. Will test and make PR.

**Comment by richard67** [comment]
@Quy PR #27228 is almost ready. I only have to provide testing instructions and a modified update package. The problem is that I have other, unrelated problems with language stings with update packages from current nightly build, which makes testing not nice.

**Comment by richard67** [comment]
@Quy PR #27228 is ready, please test. And if it works, maybe you can modify the update zip containter I provided in some comment and my test result above by that change and give this PR here a good test, too?

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on d7c03e75de187ae66b7b624c76177a02a9498499<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27032">issues.joomla.org/tracker/joomla-cms/27032</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27032">issues.joomla.org/tracker/joomla-cms/27032</a>.</sub>

**Comment by HLeithner** [comment]
@wilsonge can we merge this? looks correct for me

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**